cpp/src/copying/gather.cuh

```diff
@@ -0,0 +1,337 @@
+   * @param gather_map An iterator over integral values representing the gather map
+   * @param destination_column The column to gather into
+   * @param ignore_out_of_bounds Ignore values in `gather_map` that are
+   * out of bounds
+   * @param stream Optional CUDA stream on which to execute kernels
+   *---------------------------------------------------------------------------**/
+  template <typename column_type, typename iterator_type,
+    std::enable_if_t<is_fixed_width<column_type>()>* = nullptr>
+    std::unique_ptr<column> operator()(column_view const& source_column,
+				       iterator_type gather_map,
+				       size_type num_destination_rows,
+				       bool ignore_out_of_bounds,
+				       cudaStream_t stream) {
+
+    std::unique_ptr<column> destination_column =
+      allocate_like(source_column, num_destination_rows);
+
+
+    column_type const *source_data{source_column.data<column_type>()};
+    column_type *destination_data{destination_column->mutable_view().data<column_type>()};
+
+    if (ignore_out_of_bounds) {
+      thrust::gather_if(rmm::exec_policy(stream)->on(stream), gather_map,
+			gather_map + num_destination_rows, gather_map,
+			source_data, destination_data,
+			bounds_checker<decltype(*gather_map)>{0, source_column.size()});
+    } else {
+      thrust::gather(rmm::exec_policy(stream)->on(stream), gather_map,
+		     gather_map+num_destination_rows, source_data,
+		     destination_data);
+    }
+
+    CHECK_STREAM(stream);
+    return destination_column;
+  }
+
+  template <typename column_type, typename iterator_type,
+    std::enable_if_t<not is_fixed_width<column_type>()>* = nullptr>
+  std::unique_ptr<column> operator()(column_view const& source_column,
+				     iterator_type gather_map,
+				     size_type num_destination_rows,
+				     bool ignore_out_of_bounds,
+				     util::cuda::scoped_stream stream) {
+    CUDF_FAIL("Column type must be numeric");
+  }
+
+};
+
+/**
+ * @brief Specifies the behavior of index_converter
+ */
+enum index_conversion {
+    NEGATIVE_TO_POSITIVE = 0,
+    NONE
+};
+
+/**---------------------------------------------------------------------------*
+* @brief Function object for applying a transformation on the gathermap
+* that converts negative indices to positive indices
+*
+* A negative index `i` is transformed to `i + size`, where `size` is
+* the number of elements in the column being gathered from.
+* Allowable values for the index `i` are in the range `[-size, size)`.
+* Thus, when gathering from a column of size `10`, the index `-1`
+* is transformed to `9` (i.e., the last element), `-2` is transformed
+* to `8` (the second-to-last element) and so on.
+* Positive indices are unchanged by this transformation.
+*---------------------------------------------------------------------------**/
+template <typename map_type, index_conversion ic=index_conversion::NONE>
+struct index_converter : public thrust::unary_function<map_type,map_type>{};
```

Comment:
I'm trying to understand the need for `index_converter`. Are there plans to add other kinds of `index_conversion`? 

Otherwise, I think it adds a layer of unneeded complexity. But maybe I'm missing something.

Reply:
Do you mean as opposed to a lambda? We could replace `index_conversion` with a bool template parameter...

Reply:
Yeah, I was trying to understand why we want to use `index_converter` instead of the lambda I suggested above.